File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/proxies/InstanceProxy.java
package com.github.ol_loginov.heaplibweb.services.proxies;

import com.github.ol_loginov.heaplibweb.repository.heap.HeapScope;
import com.github.ol_loginov.heaplibweb.repository.heap.InstanceEntity;
import com.github.ol_loginov.heaplibweb.repository.heap.JavaClassEntity;
import lombok.RequiredArgsConstructor;
import org.netbeans.lib.profiler.heap.FieldValue;
import org.netbeans.lib.profiler.heap.Instance;
import org.netbeans.lib.profiler.heap.JavaClass;
import org.netbeans.lib.profiler.heap.Value;

import java.util.List;

@RequiredArgsConstructor
public class InstanceProxy implements Instance {
	private final InstanceEntity entity;
	private final HeapScope scope;

	public static Instance wrap(InstanceEntity entity, HeapScope scope) {
		return new InstanceProxy(entity, scope);
	}

	@Override
	public List<FieldValue> getFieldValues() {
		return scope.getFieldValues().streamInstanceFieldValues(entity.getInstanceId())
			.map(e -> FieldValueProxy.wrap(entity.getHeapId(), e, scope))
			.toList();
	}

	@Override
	public boolean isGCRoot() {
		return entity.isGcRoot();
	}

	@Override
	public long getInstanceId() {
		return entity.getInstanceId();
	}

	@Override
	public int getInstanceNumber() {
		return HeapScope.shouldBeReady(entity.getInstanceNumber());
	}

	@Override
	public JavaClass getJavaClass() {
		return scope.getJavaClasses().findById(new JavaClassEntity.PK(entity.getHeapId(), entity.getJavaClassId()))
			.map(e -> (JavaClass) new JavaClassProxy(e, scope))
			.orElse(null);
	}

	@Override
	public long getSize() {
		return entity.getSize();
	}

	@Override
	public long getReachableSize() {
		return HeapScope.shouldBeReady(entity.getReachableSize());
	}

	@Override
	public long getRetainedSize() {
		return HeapScope.shouldBeReady(entity.getRetainedSize());
	}

	@Override
	public Object getValueOfField(String name) {
		var fieldValueEntity = scope.getFieldValues()
			.findOneByInstanceAndFieldName(entity.getInstanceId(), name)
			.orElse(null);
		if (fieldValueEntity == null) {
			return null;
		}
		return FieldValueProxy.getValueObject(fieldValueEntity, entity.getHeapId(), scope);
	}

	@Override
	public Instance getNearestGCRootPointer() {
		throw new UnsupportedOperationException();
	}

	@Override
	public List<Value> getReferences() {
		throw new UnsupportedOperationException();
	}

	@Override
	public List<FieldValue> getStaticFieldValues() {
		throw new UnsupportedOperationException();
	}

	@Override
	public String toString() {
		return getClass().getSimpleName() + "#" + entity.getInstanceId();
	}
}

File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/proxies/InstanceProxy.kt
package com.github.ol_loginov.heaplibweb.services.proxies

import com.github.ol_loginov.heaplibweb.repository.heap.HeapRepositories
import com.github.ol_loginov.heaplibweb.repository.heap.InstanceEntity
import org.netbeans.lib.profiler.heap.FieldValue
import org.netbeans.lib.profiler.heap.Instance
import org.netbeans.lib.profiler.heap.JavaClass
import org.netbeans.lib.profiler.heap.Value

class InstanceProxy(
    private val entity: InstanceEntity,
    private val heapRepositories: HeapRepositories
) : Instance {
    companion object {
        @JvmStatic
        fun wrap(entity: InstanceEntity, heapRepositories: HeapRepositories): Instance = InstanceProxy(entity, heapRepositories)
    }

    override fun getFieldValues(): List<FieldValue> = heapRepositories
        .fieldValues.streamInstanceFieldValues(entity.instanceId)
        .map { FieldValueProxy.wrap(it, heapRepositories) }
        .toList()

    override fun isGCRoot(): Boolean = entity.rootTag > 0
    override fun getInstanceId(): Long = entity.instanceId
    override fun getInstanceNumber(): Int = entity.instanceNumber

    override fun getJavaClass(): JavaClass? = heapRepositories
        .classes.findById(entity.javaClassId)
        ?.let { JavaClassProxy(it, heapRepositories) }

    override fun getSize(): Long = entity.size.toLong()
    override fun getReachableSize(): Long = entity.reachableSize ?: throw UnsupportedOperationException("number not ready")
    override fun getRetainedSize(): Long = entity.retainedSize ?: throw UnsupportedOperationException("number not ready")

    override fun getValueOfField(name: String): Any? {
        val fieldValueEntity = heapRepositories.fieldValues
            .findOneByInstanceAndFieldName(entity.instanceId, name)
            ?: return null
        return FieldValueProxy.getValueObject(fieldValueEntity, heapRepositories)
    }

    override fun getNearestGCRootPointer(): Instance = throw UnsupportedOperationException()
    override fun getReferences(): List<Value> = throw UnsupportedOperationException()
    override fun getStaticFieldValues(): List<FieldValue> = throw UnsupportedOperationException()

    override fun toString(): String = "${this::class.java.simpleName}#${entity.instanceId}"
}
